File: bindings/java/src/main/java/io/angzarr/router/AggregateDispatch.java
package io.angzarr.router;

import io.angzarr.router.Thunks.CommandThunk;
import io.angzarr.router.Thunks.RejectionThunk;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * One aggregate component's registration: its name, domain, rebuilder, command
 * handlers, and ordered rejection compensators. The shape mirrors the engine's
 * so generated wiring targets it with minimal emitter changes.
 */
public final class AggregateDispatch {
  final String name;
  final String domain;
  final Rebuilder rebuilder;
  final Map<String, CommandThunk> commands = new LinkedHashMap<>();
  final Map<String, List<RejectionThunk>> rejections = new LinkedHashMap<>();

  public AggregateDispatch(String name, String domain, Rebuilder rebuilder) {
    this.name = name;
    this.domain = domain;
    this.rebuilder = rebuilder;
  }

  /** Registers a handler for one fully-qualified command type. */
  public AggregateDispatch onCommand(String fullName, CommandThunk thunk) {
    commands.put(fullName, thunk);
    return this;
  }

  /** Appends a compensator for one fully-qualified command type; repeated calls
   * register an ordered fan-out. */
  public AggregateDispatch onRejected(String fqCommand, RejectionThunk thunk) {
    rejections.computeIfAbsent(fqCommand, k -> new ArrayList<>()).add(thunk);
    return this;
  }
}

File: bindings/csharp/src/AggregateDispatch.cs
using System.Collections.Generic;
using Google.Protobuf;

namespace Angzarr.Router;

/// <summary>
/// One aggregate component's registration: its name, domain, rebuilder, command
/// handlers, and ordered rejection compensators. Generic in the state message so
/// handler thunks see the concrete state — the generated wiring is cast-free.
/// </summary>
public sealed class AggregateDispatch<TState>
    where TState : class, IMessage
{
    internal readonly string Name;
    internal readonly string Domain;
    internal readonly Rebuilder<TState> Rebuilder;
    internal readonly Dictionary<string, CommandThunk<TState>> Commands = new();
    internal readonly Dictionary<string, List<RejectionThunk<TState>>> Rejections = new();

    public AggregateDispatch(string name, string domain, Rebuilder<TState> rebuilder)
    {
        Name = name;
        Domain = domain;
        Rebuilder = rebuilder;
    }

    /// <summary>Registers a handler for one fully-qualified command type.</summary>
    public AggregateDispatch<TState> OnCommand(string fullName, CommandThunk<TState> thunk)
    {
        Commands[fullName] = thunk;
        return this;
    }

    /// <summary>Appends a compensator for one fully-qualified command type;
    /// repeated calls register an ordered fan-out.</summary>
    public AggregateDispatch<TState> OnRejected(string fqCommand, RejectionThunk<TState> thunk)
    {
        if (!Rejections.TryGetValue(fqCommand, out var list))
        {
            list = new List<RejectionThunk<TState>>();
            Rejections[fqCommand] = list;
        }
        list.Add(thunk);
        return this;
    }
}
